src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,467 @@
+	}
+	else if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_PREFER_DISTRIBUTED)
+	{
+		return distributedRTECandidate ? distributedRTECandidate : localRTECandidate;
+	}
+	else
+	{
+		/*
+		 * We want to convert distributed tables only if all the distributed tables
+		 * have a constant filter on a unique index, otherwise we would be redundantly
+		 * converting a distributed table as we will convert all the other local tables.
+		 */
+		bool allRangeTableEntriesHaveUniqueIndex = AllRangeTableEntriesHaveUniqueIndex(
+			conversionCandidates->distributedTableList);
+
+		if (allRangeTableEntriesHaveUniqueIndex)
+		{
+			return distributedRTECandidate ? distributedRTECandidate : localRTECandidate;
+		}
+		else
+		{
+			return localRTECandidate ? localRTECandidate : distributedRTECandidate;
+		}
+	}
+}
+
+
+/*
+ * AllRangeTableEntriesHaveUniqueIndex returns true if all of the RTE's in the given
+ * list have a unique index.
+ */
+static bool
+AllRangeTableEntriesHaveUniqueIndex(List *rangeTableEntryDetailsList)
+{
+	RangeTableEntryDetails *rangeTableEntryDetails = NULL;
+	foreach_ptr(rangeTableEntryDetails, rangeTableEntryDetailsList)
+	{
+		if (!rangeTableEntryDetails->hasConstantFilterOnUniqueColumn)
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * RemoveFromConversionCandidates removes an element from
+ * the relevant list based on the relation id.
+ */
+static void
+RemoveFromConversionCandidates(ConversionCandidates *conversionCandidates, int
+							   rteIdentity)
+{
+	RangeTableEntryDetails *rangeTableEntryDetails = NULL;
+	foreach_ptr(rangeTableEntryDetails, conversionCandidates->localTableList)
+	{
+		if (rangeTableEntryDetails->rteIdentity == rteIdentity)
+		{
+			conversionCandidates->localTableList =
+				list_delete_ptr(conversionCandidates->localTableList,
+								rangeTableEntryDetails);
+			return;
+		}
+	}
+
+	foreach_ptr(rangeTableEntryDetails, conversionCandidates->distributedTableList)
+	{
+		if (rangeTableEntryDetails->rteIdentity == rteIdentity)
+		{
+			conversionCandidates->distributedTableList =
+				list_delete_ptr(conversionCandidates->distributedTableList,
+								rangeTableEntryDetails);
+			return;
+		}
+	}
+
+	ereport(ERROR, (errmsg("invalid rte index is given :%d", rteIdentity)));
+}
+
+
+/*
+ * ShouldConvertLocalTableJoinsToSubqueries returns true if we should
+ * convert local-dist table joins to subqueries.
+ */
+bool
+ShouldConvertLocalTableJoinsToSubqueries(Query *query, List *rangeTableList,
+										 PlannerRestrictionContext *
+										 plannerRestrictionContext)
+{
+	if (LocalTableJoinPolicy == LOCAL_JOIN_POLICY_NEVER)
+	{
+		/* user doesn't want Citus to enable local table joins */
+		return false;
+	}
+
+	if (!ContainsTableToBeConvertedToSubquery(rangeTableList))
+	{
+		return false;
+	}
+
+	plannerRestrictionContext = FilterPlannerRestrictionForQuery(
+		plannerRestrictionContext, query);
+	if (IsRouterPlannable(query, plannerRestrictionContext))
```

Comment:
I guess my only high-level concern with this PR is the expensiveness of this. Do you see a path towards speeding this up in the future?

Reply:
At a high level as we are trying to convert this query to a router plannable query, I guess we would need to have a function like this here, but we probably don't need to perform the same passed checks again and again in this function. If we passed some check in this function before a conversion, I guess it won't fail after a conversion either, so I wonder if we could structure the checks in this function (also in RouterJob) in a way that we also know what checks passed, as in what error we got last time, and continue from there.

Reply:
Just as a reference:
Planning of 1 distributed vs 25 local tables take between 300-500 ms. 
Planning of 25 local tables take 15-40 ms
Planning of 1 distributed vs 5 local tables take around 5 ms
Planning of 5 local tables take around 1ms